src/csharp/Grpc.Core.Api/AsyncCallState.cs

```diff
@@ -64,21 +64,21 @@ internal struct AsyncCallState
         internal Task<Metadata> ResponseHeadersAsync()
         {
             var withState = responseHeadersAsync as Func<object, Task<Metadata>>;
-            return withState != null ? withState(callbackState)
+            return withState != null ? withState(callbackState!)
```

Comment:
```suggestion
            return withState != null && callbackState != null ? withState(callbackState)
```

Reply:
If withState is not null then callbackState has to also be not null. I chose not to use a Debug.Assert because this library targets netstandard2.0 and Debug.Assert isn't annotated there to correctly suppress the nullable warning.

Reply:
Makes sense. It would be neat if we could use [MemberNotNullWhen](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.membernotnullwhenattribute?view=net-5.0) but I'm not sure if it can be used in private members.